File: apps/api/internal/modules/production/application/production_plan_service_test.go
package application

import (
	"context"
	"errors"
	"testing"
	"time"

	inventoryapp "github.com/Chinsusu/ERP-v2/apps/api/internal/modules/inventory/application"
	inventorydomain "github.com/Chinsusu/ERP-v2/apps/api/internal/modules/inventory/domain"
	masterdatadomain "github.com/Chinsusu/ERP-v2/apps/api/internal/modules/masterdata/domain"
	productiondomain "github.com/Chinsusu/ERP-v2/apps/api/internal/modules/production/domain"
	"github.com/Chinsusu/ERP-v2/apps/api/internal/shared/audit"
	"github.com/Chinsusu/ERP-v2/apps/api/internal/shared/decimal"
)

func TestProductionPlanServiceCreatesFormulaSnapshotDemandAndPurchaseRequestDraft(t *testing.T) {
	ctx := context.Background()
	now := time.Date(2026, 5, 4, 10, 0, 0, 0, time.UTC)
	formula := activeProductionPlanFormula(t)
	store := NewPrototypeProductionPlanStore(audit.NewInMemoryLogStore())
	service := NewProductionPlanService(
		store,
		fakeProductionPlanFormulaReader{formula: formula},
		fakeProductionPlanAvailableStock{
			rows: []inventorydomain.AvailableStockSnapshot{
				{
					ItemID:       "item-act-baicapil",
					SKU:          "ACT_BAICAPIL",
					BaseUOMCode:  decimal.MustUOMCode("KG"),
					AvailableQty: decimal.MustQuantity("0.000500"),
				},
			},
		},
	)
	service.clock = func() time.Time { return now }

	result, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:               "plan-001",
		PlanNo:           "PP-260504-0001",
		OutputItemID:     "item-xff-150",
		PlannedQty:       "162",
		UOMCode:          "PCS",
		PlannedStartDate: "2026-05-10",
		PlannedEndDate:   "2026-05-11",
		ActorID:          "user-production",
		RequestID:        "req-production-plan",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}

	plan := result.ProductionPlan
	if plan.FormulaID != formula.ID || plan.FormulaVersion != "v1" || len(plan.Lines) != 1 {
		t.Fatalf("plan snapshot = %+v, want active formula snapshot with one line", plan)
	}
	line := plan.Lines[0]
	if line.RequiredStockBaseQty != "0.162000" || line.AvailableQty != "0.000500" || line.ShortageQty != "0.161500" {
		t.Fatalf("line = %+v, want shortage from formula demand minus available stock", line)
	}
	if len(plan.PurchaseDraft.Lines) != 1 {
		t.Fatalf("purchase draft lines = %d, want 1", len(plan.PurchaseDraft.Lines))
	}
	prLine := plan.PurchaseDraft.Lines[0]
	if prLine.SKU != "ACT_BAICAPIL" || prLine.RequestedQty != "0.161500" || prLine.UOMCode != "KG" {
		t.Fatalf("purchase draft line = %+v, want shortage draft only", prLine)
	}
	if result.AuditLogID == "" {
		t.Fatalf("audit id is empty")
	}
	if store.Count() != 1 {
		t.Fatalf("stored plans = %d, want 1", store.Count())
	}
}

func TestProductionPlanServiceDoesNotCreatePurchaseDraftWhenEnoughStock(t *testing.T) {
	ctx := context.Background()
	formula := activeProductionPlanFormula(t)
	store := NewPrototypeProductionPlanStore(audit.NewInMemoryLogStore())
	service := NewProductionPlanService(
		store,
		fakeProductionPlanFormulaReader{formula: formula},
		fakeProductionPlanAvailableStock{
			rows: []inventorydomain.AvailableStockSnapshot{
				{
					ItemID:       "item-act-baicapil",
					SKU:          "ACT_BAICAPIL",
					BaseUOMCode:  decimal.MustUOMCode("KG"),
					AvailableQty: decimal.MustQuantity("1.000000"),
				},
			},
		},
	)

	result, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-002",
		PlanNo:       "PP-260504-0002",
		OutputItemID: "item-xff-150",
		PlannedQty:   "162",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}
	if len(result.ProductionPlan.PurchaseDraft.Lines) != 0 {
		t.Fatalf("purchase draft = %+v, want no draft when stock is enough", result.ProductionPlan.PurchaseDraft)
	}
	if result.ProductionPlan.Lines[0].NeedsPurchase || result.ProductionPlan.Lines[0].ShortageQty != "0.000000" {
		t.Fatalf("line = %+v, want no shortage", result.ProductionPlan.Lines[0])
	}
}

func TestProductionPlanServiceCreatesWarehouseIssueFromReadyDemandLineAndRollsUpPostedIssue(t *testing.T) {
	ctx := context.Background()
	formula := activeProductionPlanFormula(t)
	auditStore := audit.NewInMemoryLogStore()
	planStore := NewPrototypeProductionPlanStore(auditStore)
	issueStore := inventoryapp.NewPrototypeWarehouseIssueStore()
	issueService := inventoryapp.NewWarehouseIssueService(
		issueStore,
		inventoryapp.NewInMemoryStockMovementStore(),
		auditStore,
	).WithClock(func() time.Time {
		return time.Date(2026, 5, 5, 11, 0, 0, 0, time.UTC)
	})
	service := NewProductionPlanService(
		planStore,
		fakeProductionPlanFormulaReader{formula: formula},
		fakeProductionPlanAvailableStock{
			rows: []inventorydomain.AvailableStockSnapshot{
				{
					WarehouseID:   "wh-hcm-rm",
					WarehouseCode: "WH-HCM-RM",
					LocationID:    "bin-rm-a01",
					LocationCode:  "RM-A01",
					ItemID:        "item-act-baicapil",
					SKU:           "ACT_BAICAPIL",
					BatchID:       "batch-act-baicapil-001",
					BatchNo:       "LOT-ACT-001",
					BaseUOMCode:   decimal.MustUOMCode("KG"),
					AvailableQty:  decimal.MustQuantity("1.000000"),
				},
			},
		},
	).WithWarehouseIssueService(issueService)

	created, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-issue-001",
		PlanNo:       "PP-260505-ISSUE-001",
		OutputItemID: "item-xff-150",
		PlannedQty:   "162",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}
	lineID := created.ProductionPlan.Lines[0].ID

	issueResult, err := service.CreateWarehouseIssueFromProductionPlan(ctx, CreateProductionPlanWarehouseIssueInput{
		PlanID:          created.ProductionPlan.ID,
		LineIDs:         []string{lineID},
		DestinationType: "factory",
		DestinationName: "Factory A",
		ReasonCode:      "production_plan_issue",
		ActorID:         "user-production",
		RequestID:       "req-plan-issue-create",
	})
	if err != nil {
		t.Fatalf("CreateWarehouseIssueFromProductionPlan() error = %v", err)
	}
	issue := issueResult.WarehouseIssue
	if issue.WarehouseID != "wh-hcm-rm" || issue.Lines[0].LocationID != "bin-rm-a01" {
		t.Fatalf("issue location = %s/%s, want selected available stock warehouse/bin", issue.WarehouseID, issue.Lines[0].LocationID)
	}
	if issue.Lines[0].SourceDocumentType != "production_plan" ||
		issue.Lines[0].SourceDocumentID != created.ProductionPlan.ID ||
		issue.Lines[0].SourceDocumentLineID != lineID {
		t.Fatalf("issue source = %+v, want production plan and demand line source", issue.Lines[0])
	}
	if issue.Lines[0].Quantity != created.ProductionPlan.Lines[0].RequiredStockBaseQty {
		t.Fatalf("issue qty = %s, want required stock base qty %s", issue.Lines[0].Quantity, created.ProductionPlan.Lines[0].RequiredStockBaseQty)
	}

	for _, action := range []string{"submit", "approve", "post"} {
		if _, err := issueService.TransitionWarehouseIssue(ctx, inventoryapp.WarehouseIssueTransitionInput{
			ID:      issue.ID,
			ActorID: "warehouse-lead",
			Action:  action,
		}); err != nil {
			t.Fatalf("%s issue: %v", action, err)
		}
	}

	updated, err := service.GetProductionPlan(ctx, created.ProductionPlan.ID)
	if err != nil {
		t.Fatalf("GetProductionPlan() error = %v", err)
	}
	updatedLine := updated.Lines[0]
	if updatedLine.IssuedQty != updatedLine.RequiredStockBaseQty ||
		updatedLine.RemainingIssueQty != "0.000000" ||
		updatedLine.IssueStatus != productiondomain.ProductionPlanIssueStatusIssued {
		t.Fatalf("updated line = %+v, want issued readiness after posted Warehouse Issue", updatedLine)
	}
	if len(updatedLine.WarehouseIssues) != 1 || updatedLine.WarehouseIssues[0].IssueNo != issue.IssueNo {
		t.Fatalf("warehouse issue refs = %+v, want posted issue ref", updatedLine.WarehouseIssues)
	}
}

func TestProductionPlanServiceRejectsWarehouseIssueFromShortageLine(t *testing.T) {
	ctx := context.Background()
	auditStore := audit.NewInMemoryLogStore()
	service := NewProductionPlanService(
		NewPrototypeProductionPlanStore(auditStore),
		fakeProductionPlanFormulaReader{formula: activeProductionPlanFormula(t)},
		fakeProductionPlanAvailableStock{
			rows: []inventorydomain.AvailableStockSnapshot{
				{
					WarehouseID:   "wh-hcm-rm",
					WarehouseCode: "WH-HCM-RM",
					ItemID:        "item-act-baicapil",
					SKU:           "ACT_BAICAPIL",
					BaseUOMCode:   decimal.MustUOMCode("KG"),
					AvailableQty:  decimal.MustQuantity("0.000500"),
				},
			},
		},
	).WithWarehouseIssueService(inventoryapp.NewWarehouseIssueService(
		inventoryapp.NewPrototypeWarehouseIssueStore(),
		inventoryapp.NewInMemoryStockMovementStore(),
		auditStore,
	))

	created, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-shortage-001",
		PlanNo:       "PP-260505-SHORTAGE-001",
		OutputItemID: "item-xff-150",
		PlannedQty:   "162",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}

	_, err = service.CreateWarehouseIssueFromProductionPlan(ctx, CreateProductionPlanWarehouseIssueInput{
		PlanID:          created.ProductionPlan.ID,
		LineIDs:         []string{created.ProductionPlan.Lines[0].ID},
		DestinationType: "factory",
		DestinationName: "Factory A",
		ReasonCode:      "production_plan_issue",
		ActorID:         "user-production",
	})
	if !errors.Is(err, ErrProductionPlanMaterialIssueNotReady) {
		t.Fatalf("CreateWarehouseIssueFromProductionPlan() error = %v, want ErrProductionPlanMaterialIssueNotReady", err)
	}
}

func TestProductionPlanServiceSubmitsApprovesAndConvertsPurchaseRequestDraft(t *testing.T) {
	ctx := context.Background()
	now := time.Date(2026, 5, 5, 9, 0, 0, 0, time.UTC)
	store := NewPrototypeProductionPlanStore(audit.NewInMemoryLogStore())
	service := newProductionPlanServiceWithShortage(t, store)
	service.clock = func() time.Time { return now }

	created, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-001",
		PlanNo:       "PP-260505-0001",
		OutputItemID: "item-xff-150",
		PlannedQty:   "162",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}
	draftID := created.ProductionPlan.PurchaseDraft.ID

	_, err = service.SubmitPurchaseRequestDraft(ctx, PurchaseRequestDraftActionInput{
		ID:             draftID,
		ExpectedStatus: productiondomain.PurchaseRequestDraftStatusSubmitted,
		ActorID:        "user-purchase",
		RequestID:      "req-pr-submit-stale",
	})
	if !errors.Is(err, productiondomain.ErrProductionPlanInvalidPurchaseRequestTransition) {
		t.Fatalf("stale submit error = %v, want ErrProductionPlanInvalidPurchaseRequestTransition", err)
	}

	submitted, err := service.SubmitPurchaseRequestDraft(ctx, PurchaseRequestDraftActionInput{
		ID:        draftID,
		ActorID:   "user-purchase",
		RequestID: "req-pr-submit",
	})
	if err != nil {
		t.Fatalf("SubmitPurchaseRequestDraft() error = %v", err)
	}
	if submitted.CurrentStatus != productiondomain.PurchaseRequestDraftStatusSubmitted {
		t.Fatalf("submitted status = %q, want submitted", submitted.CurrentStatus)
	}
	service.clock = func() time.Time { return now.Add(time.Hour) }
	approved, err := service.ApprovePurchaseRequestDraft(ctx, PurchaseRequestDraftActionInput{
		ID:        draftID,
		ActorID:   "user-manager",
		RequestID: "req-pr-approve",
	})
	if err != nil {
		t.Fatalf("ApprovePurchaseRequestDraft() error = %v", err)
	}
	if approved.CurrentStatus != productiondomain.PurchaseRequestDraftStatusApproved {
		t.Fatalf("approved status = %q, want approved", approved.CurrentStatus)
	}
	service.clock = func() time.Time { return now.Add(2 * time.Hour) }
	converted, err := service.MarkPurchaseRequestDraftConverted(ctx, ConvertPurchaseRequestDraftInput{
		ID:              draftID,
		PurchaseOrderID: "po-260505-0001",
		PurchaseOrderNo: "PO-260505-0001",
		ActorID:         "user-purchase",
		RequestID:       "req-pr-convert",
	})
	if err != nil {
		t.Fatalf("MarkPurchaseRequestDraftConverted() error = %v", err)
	}
	if converted.PurchaseRequestDraft.Status != productiondomain.PurchaseRequestDraftStatusConvertedToPO {
		t.Fatalf("converted status = %q, want converted_to_po", converted.PurchaseRequestDraft.Status)
	}
	if converted.PurchaseRequestDraft.ConvertedPurchaseOrderID != "po-260505-0001" {
		t.Fatalf("converted PO id = %q, want po-260505-0001", converted.PurchaseRequestDraft.ConvertedPurchaseOrderID)
	}

	drafts, err := service.ListPurchaseRequestDrafts(ctx, PurchaseRequestDraftFilter{Search: "PP-260505"})
	if err != nil {
		t.Fatalf("ListPurchaseRequestDrafts() error = %v", err)
	}
	if len(drafts) != 1 || drafts[0].Status != productiondomain.PurchaseRequestDraftStatusConvertedToPO {
		t.Fatalf("drafts = %+v, want one converted draft", drafts)
	}
}

func TestProductionPlanServiceRejectsPurchaseRequestApprovalBeforeSubmit(t *testing.T) {
	ctx := context.Background()
	store := NewPrototypeProductionPlanStore(audit.NewInMemoryLogStore())
	service := newProductionPlanServiceWithShortage(t, store)
	created, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-001",
		PlanNo:       "PP-260505-0001",
		OutputItemID: "item-xff-150",
		PlannedQty:   "162",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}

	_, err = service.ApprovePurchaseRequestDraft(ctx, PurchaseRequestDraftActionInput{
		ID:      created.ProductionPlan.PurchaseDraft.ID,
		ActorID: "user-manager",
	})
	if !errors.Is(err, productiondomain.ErrProductionPlanInvalidPurchaseRequestTransition) {
		t.Fatalf("error = %v, want ErrProductionPlanInvalidPurchaseRequestTransition", err)
	}
}

func TestProductionPlanServiceAcceptsPublicItemReferenceResolvedByFormulaList(t *testing.T) {
	ctx := context.Background()
	formula := activeProductionPlanFormula(t)
	formula.FinishedItemID = "2e2f71b4-a502-43e8-a448-04d875a04cb5"
	store := NewPrototypeProductionPlanStore(audit.NewInMemoryLogStore())
	service := NewProductionPlanService(
		store,
		fakeProductionPlanFormulaReader{formula: formula, outputAliases: []string{"item-xff-150"}},
		fakeProductionPlanAvailableStock{},
	)

	result, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-public-ref-list",
		PlanNo:       "PP-260504-PUBLIC-LIST",
		OutputItemID: "item-xff-150",
		PlannedQty:   "10",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}
	if result.ProductionPlan.OutputItemID != formula.FinishedItemID {
		t.Fatalf("OutputItemID = %q, want formula finished item UUID %q", result.ProductionPlan.OutputItemID, formula.FinishedItemID)
	}
}

func newProductionPlanServiceWithShortage(t *testing.T, store *PrototypeProductionPlanStore) ProductionPlanService {
	t.Helper()
	return NewProductionPlanService(
		store,
		fakeProductionPlanFormulaReader{formula: activeProductionPlanFormula(t)},
		fakeProductionPlanAvailableStock{
			rows: []inventorydomain.AvailableStockSnapshot{
				{
					ItemID:       "item-act-baicapil",
					SKU:          "ACT_BAICAPIL",
					BaseUOMCode:  decimal.MustUOMCode("KG"),
					AvailableQty: decimal.MustQuantity("0.000500"),
				},
			},
		},
	)
}

func TestProductionPlanServiceAcceptsPublicItemReferenceWithExplicitFormulaID(t *testing.T) {
	ctx := context.Background()
	formula := activeProductionPlanFormula(t)
	formula.FinishedItemID = "2e2f71b4-a502-43e8-a448-04d875a04cb5"
	store := NewPrototypeProductionPlanStore(audit.NewInMemoryLogStore())
	service := NewProductionPlanService(
		store,
		fakeProductionPlanFormulaReader{formula: formula, outputAliases: []string{"item-xff-150"}},
		fakeProductionPlanAvailableStock{},
	)

	result, err := service.CreateProductionPlan(ctx, CreateProductionPlanInput{
		ID:           "plan-public-ref-get",
		PlanNo:       "PP-260504-PUBLIC-GET",
		OutputItemID: "item-xff-150",
		FormulaID:    formula.ID,
		PlannedQty:   "10",
		UOMCode:      "PCS",
		ActorID:      "user-production",
	})
	if err != nil {
		t.Fatalf("CreateProductionPlan() error = %v", err)
	}
	if result.ProductionPlan.FormulaID != formula.ID {
		t.Fatalf("FormulaID = %q, want %q", result.ProductionPlan.FormulaID, formula.ID)
	}
}

func activeProductionPlanFormula(t *testing.T) masterdatadomain.Formula {
	t.Helper()
	formula, err := masterdatadomain.NewFormula(masterdatadomain.NewFormulaInput{
		ID:               "formula-xff-v1",
		FormulaCode:      "XFF-150ML",
		FinishedItemID:   "item-xff-150",
		FinishedSKU:      "XFF",
		FinishedItemName: "Tinh chat buoi Fast & Furious 150ML",
		FinishedItemType: masterdatadomain.ItemTypeFinishedGood,
		FormulaVersion:   "v1",
		BatchQty:         decimal.MustQuantity("81"),
		BatchUOMCode:     "PCS",
		BaseBatchQty:     decimal.MustQuantity("81"),
		BaseBatchUOMCode: "PCS",
		Status:           masterdatadomain.FormulaStatusActive,
		ApprovalStatus:   masterdatadomain.FormulaApprovalApproved,
		Lines: []masterdatadomain.NewFormulaLineInput{
			{
				ID:               "formula-line-001",
				LineNo:           1,
				ComponentItemID:  "item-act-baicapil",
				ComponentSKU:     "ACT_BAICAPIL",
				ComponentName:    "BAICAPIL",
				ComponentType:    masterdatadomain.FormulaComponentRawMaterial,
				EnteredQty:       decimal.MustQuantity("0.001"),
				EnteredUOMCode:   "KG",
				CalcQty:          decimal.MustQuantity("1"),
				CalcUOMCode:      "G",
				StockBaseQty:     decimal.MustQuantity("0.001"),
				StockBaseUOMCode: "KG",
				WastePercent:     decimal.MustRate("0"),
				IsRequired:       true,
				IsStockManaged:   true,
				LineStatus:       masterdatadomain.FormulaLineStatusActive,
			},
		},
	})
	if err != nil {
		t.Fatalf("formula fixture: %v", err)
	}

	return formula
}

type fakeProductionPlanFormulaReader struct {
	formula       masterdatadomain.Formula
	err           error
	outputAliases []string
}

func (r fakeProductionPlanFormulaReader) Get(ctx context.Context, id string) (masterdatadomain.Formula, error) {
	if r.err != nil {
		return masterdatadomain.Formula{}, r.err
	}
	if id != "" && id != r.formula.ID {
		return masterdatadomain.Formula{}, errors.New("formula not found")
	}

	return r.formula.Clone(), nil
}

func (r fakeProductionPlanFormulaReader) List(ctx context.Context, filter masterdatadomain.FormulaFilter) ([]masterdatadomain.Formula, error) {
	if r.err != nil {
		return nil, r.err
	}
	if filter.FinishedItemID != "" && filter.FinishedItemID != r.formula.FinishedItemID && !r.matchesOutputAlias(filter.FinishedItemID) {
		return nil, nil
	}
	if filter.Status != "" && filter.Status != r.formula.Status {
		return nil, nil
	}

	return []masterdatadomain.Formula{r.formula.Clone()}, nil
}

func (r fakeProductionPlanFormulaReader) matchesOutputAlias(outputItemID string) bool {
	for _, alias := range r.outputAliases {
		if alias == outputItemID {
			return true
		}
	}

	return false
}

type fakeProductionPlanAvailableStock struct {
	rows []inventorydomain.AvailableStockSnapshot
}

func (s fakeProductionPlanAvailableStock) Execute(ctx context.Context, filter inventorydomain.AvailableStockFilter) ([]inventorydomain.AvailableStockSnapshot, error) {
	matches := make([]inventorydomain.AvailableStockSnapshot, 0, len(s.rows))
	for _, row := range s.rows {
		if filter.SKU != "" && row.SKU != filter.SKU {
			continue
		}
		matches = append(matches, row)
	}

	return matches, nil
}

var _ ProductionPlanStore = (*PrototypeProductionPlanStore)(nil)

File: apps/web/src/modules/production-planning/services/productionPlanService.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import {
  createProductionPlan,
  createProductionPlans,
  createWarehouseIssueFromProductionPlan,
  formatProductionPlanQuantity,
  getProductionPlan,
  summarizeProductionPlans
} from "./productionPlanService";

describe("productionPlanService", () => {
  afterEach(() => {
    vi.unstubAllGlobals();
  });

  it("creates a production plan and maps material shortages into an internal purchase request draft", async () => {
    const fetchMock = vi.fn(async () =>
      new Response(
        JSON.stringify({
          success: true,
          request_id: "req-production-plan",
          data: {
            id: "plan-001",
            org_id: "org-my-pham",
            plan_no: "PP-260504-0001",
            output_item_id: "item-xff-150",
            output_sku: "XFF",
            output_item_name: "Tinh chat buoi Fast & Furious 150ML",
            output_item_type: "finished_good",
            planned_qty: "162.000000",
            uom_code: "PCS",
            formula_id: "formula-xff-v1",
            formula_code: "XFF-150ML",
            formula_version: "v1",
            formula_batch_qty: "81.000000",
            formula_batch_uom_code: "PCS",
            status: "purchase_request_draft_created",
            lines: [
              {
                id: "pp-line-plan-001-001",
                formula_line_id: "formula-line-001",
                line_no: 1,
                component_item_id: "item-act-baicapil",
                component_sku: "ACT_BAICAPIL",
                component_name: "BAICAPIL",
                component_type: "raw_material",
                formula_qty: "1.000000",
                formula_uom_code: "G",
                required_qty: "162.000000",
                required_uom_code: "G",
                required_stock_base_qty: "0.162000",
                stock_base_uom_code: "KG",
                available_qty: "0.000500",
                shortage_qty: "0.161500",
                purchase_draft_qty: "0.161500",
                purchase_draft_uom_code: "KG",
                is_stock_managed: true,
                needs_purchase: true
              }
            ],
            purchase_request_draft: {
              id: "pr-draft-001",
              request_no: "PR-DRAFT-260504-0001",
              source_production_plan_id: "plan-001",
              source_production_plan_no: "PP-260504-0001",
              status: "draft",
              lines: [
                {
                  id: "pr-line-001",
                  line_no: 1,
                  source_production_plan_line_id: "pp-line-plan-001-001",
                  item_id: "item-act-baicapil",
                  sku: "ACT_BAICAPIL",
                  item_name: "BAICAPIL",
                  requested_qty: "0.161500",
                  uom_code: "KG"
                }
              ],
              created_at: "2026-05-04T03:00:00Z",
              created_by: "user-production"
            },
            audit_log_id: "audit-production-plan-001",
            created_at: "2026-05-04T03:00:00Z",
            created_by: "user-production",
            updated_at: "2026-05-04T03:00:00Z",
            updated_by: "user-production",
            version: 1
          }
        }),
        { status: 201, headers: { "Content-Type": "application/json" } }
      )
    );
    vi.stubGlobal("fetch", fetchMock);

    const plan = await createProductionPlan({
      outputItemId: "item-xff-150",
      formulaId: "formula-xff-v1",
      plannedQty: "162",
      uomCode: "pcs",
      plannedStartDate: "2026-05-10",
      plannedEndDate: "2026-05-11"
    });

    expect(fetchMock).toHaveBeenCalledWith(
      expect.stringContaining("/production-plans"),
      expect.objectContaining({
        method: "POST",
        body: expect.stringContaining('"planned_qty":"162.000000"')
      })
    );
    expect(plan.lines[0]).toMatchObject({
      componentSku: "ACT_BAICAPIL",
      requiredStockBaseQty: "0.162000",
      shortageQty: "0.161500",
      needsPurchase: true
    });
    expect(plan.purchaseRequestDraft.lines[0]).toMatchObject({
      sku: "ACT_BAICAPIL",
      requestedQty: "0.161500",
      uomCode: "KG"
    });
    expect(summarizeProductionPlans([plan])).toEqual({
      total: 1,
      draft: 0,
      shortageLines: 1,
      purchaseDraftLines: 1
    });
  });

  it("formats small metric quantities for readable material demand review", () => {
    expect(formatProductionPlanQuantity("0.000001", "KG")).toBe("1 mg");
    expect(formatProductionPlanQuantity("0.001500", "KG")).toBe("1,5 g");
    expect(formatProductionPlanQuantity("2.000000", "PCS")).toBe("2 PCS");
  });

  it("loads production plan detail by id", async () => {
    const fetchMock = vi.fn(async () =>
      new Response(
        JSON.stringify({
          success: true,
          request_id: "req-production-plan-detail",
          data: {
            id: "plan-001",
            org_id: "org-my-pham",
            plan_no: "PP-260504-0001",
            output_item_id: "item-aah",
            output_sku: "AAH",
            output_item_name: "Kem u phuc hoi AS A HABIT BIO 350GR",
            output_item_type: "finished_good",
            planned_qty: "999.000000",
            uom_code: "PCS",
            formula_id: "formula-aah-v1",
            formula_code: "S23SMK260504200049",
            formula_version: "v260504200049",
            formula_batch_qty: "1.000000",
            formula_batch_uom_code: "PCS",
            status: "purchase_request_draft_created",
            lines: [],
            purchase_request_draft: { lines: [] },
            created_at: "2026-05-04T03:00:00Z",
            created_by: "user-production",
            updated_at: "2026-05-04T03:00:00Z",
            updated_by: "user-production",
            version: 1
          }
        }),
        { status: 200, headers: { "Content-Type": "application/json" } }
      )
    );
    vi.stubGlobal("fetch", fetchMock);

    const plan = await getProductionPlan("plan-001");

    expect(fetchMock).toHaveBeenCalledWith(
      expect.stringContaining("/production-plans/plan-001"),
      expect.any(Object)
    );
    expect(plan.planNo).toBe("PP-260504-0001");
  });

  it("creates a source-linked warehouse issue from a production plan line", async () => {
    const fetchMock = vi.fn(async (_url, init) => {
      expect(String(_url)).toContain("/production-plans/plan-001/warehouse-issues");
      expect(JSON.parse(String((init as RequestInit).body))).toMatchObject({
        line_ids: ["pp-line-plan-001-001"],
        destination_type: "factory",
        destination_name: "Factory",
        reason_code: "production_plan_issue"
      });

      return new Response(
        JSON.stringify({
          success: true,
          request_id: "req-warehouse-issue",
          data: {
            id: "issue-001",
            issue_no: "WI-260505-0001",
            org_id: "org-my-pham",
            warehouse_id: "wh-main",
            warehouse_code: "MAIN",
            destination_type: "factory",
            destination_name: "Factory",
            reason_code: "production_plan_issue",
            status: "draft",
            requested_by: "user-production",
            lines: [
              {
                id: "issue-line-001",
                item_id: "item-aci-bha",
                sku: "ACI_BHA",
                item_name: "ACID SALICYLIC",
                quantity: "0.099900",
                base_uom_code: "KG",
                source_document_type: "production_plan",
                source_document_id: "plan-001",
                source_document_line_id: "pp-line-plan-001-001"
              }
            ],
            created_at: "2026-05-05T03:00:00Z",
            updated_at: "2026-05-05T03:00:00Z"
          }
        }),
        { status: 201, headers: { "Content-Type": "application/json" } }
      );
    });
    vi.stubGlobal("fetch", fetchMock);

    const issue = await createWarehouseIssueFromProductionPlan("plan-001", {
      lineIds: ["pp-line-plan-001-001"]
    });

    expect(issue.issueNo).toBe("WI-260505-0001");
    expect(issue.lines[0]).toMatchObject({
      sourceDocumentType: "production_plan",
      sourceDocumentId: "plan-001",
      sourceDocumentLineId: "pp-line-plan-001-001"
    });
  });

  it("creates multiple production plans from one submission", async () => {
    const fetchMock = vi.fn(async (_url, init) => {
      const body = JSON.parse(String((init as RequestInit).body));
      const sku = body.output_item_id === "item-aah" ? "AAH" : "XFF";

      return new Response(
        JSON.stringify({
          success: true,
          request_id: `req-${sku.toLowerCase()}`,
          data: {
            id: `plan-${sku.toLowerCase()}`,
            org_id: "org-my-pham",
            plan_no: `PP-260504-${sku}`,
            output_item_id: body.output_item_id,
            output_sku: sku,
            output_item_name: sku,
            output_item_type: "finished_good",
            planned_qty: body.planned_qty,
            uom_code: body.uom_code,
            formula_id: body.formula_id,
            formula_code: sku,
            formula_version: "v1",
            formula_batch_qty: "1.000000",
            formula_batch_uom_code: body.uom_code,
            status: "purchase_request_draft_created",
            lines: [],
            purchase_request_draft: { lines: [] },
            created_at: "2026-05-04T03:00:00Z",
            created_by: "user-production",
            updated_at: "2026-05-04T03:00:00Z",
            updated_by: "user-production",
            version: 1
          }
        }),
        { status: 201, headers: { "Content-Type": "application/json" } }
      );
    });
    vi.stubGlobal("fetch", fetchMock);

    const plans = await createProductionPlans([
      {
        outputItemId: "item-aah",
        formulaId: "formula-aah-v1",
        plannedQty: "10",
        uomCode: "pcs"
      },
      {
        outputItemId: "item-xff",
        formulaId: "formula-xff-v1",
        plannedQty: "50",
        uomCode: "pcs"
      }
    ]);

    expect(fetchMock).toHaveBeenCalledTimes(2);
    expect(fetchMock.mock.calls[0][1]).toEqual(
      expect.objectContaining({
        body: expect.stringContaining('"output_item_id":"item-aah"')
      })
    );
    expect(fetchMock.mock.calls[1][1]).toEqual(
      expect.objectContaining({
        body: expect.stringContaining('"output_item_id":"item-xff"')
      })
    );
    expect(plans.map((plan) => plan.outputSku)).toEqual(["AAH", "XFF"]);
  });

  it("rejects empty multi-plan submissions before calling the API", async () => {
    const fetchMock = vi.fn();
    vi.stubGlobal("fetch", fetchMock);

    await expect(createProductionPlans([])).rejects.toThrow("At least one production plan line is required");
    expect(fetchMock).not.toHaveBeenCalled();
  });
});
